linkerd/app/outbound/src/opaq/concrete.rs

```diff
@@ -1,109 +1,301 @@
-use super::{Concrete, Endpoint};
-use crate::{endpoint, stack_labels, Outbound};
+use crate::{stack_labels, Outbound};
 use linkerd_app_core::{
-    drain, io,
+    drain, io, metrics, profiles,
     proxy::{
         api_resolve::{ConcreteAddr, Metadata},
         core::Resolve,
-        tcp,
+        http::AuthorityOverride,
+        tcp::{self, balance},
     },
-    svc, Error,
+    svc::{self, layer::Layer},
+    tls,
+    transport::{self, addrs::*},
+    transport_header::SessionProtocol,
+    Error, Infallible, NameAddr,
 };
-use std::time;
+use std::{fmt::Debug, net::SocketAddr};
 use tracing::info_span;
 
+/// Parameter configuring dispatcher behavior.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum Dispatch {
+    Balance(NameAddr, balance::EwmaConfig),
+    Forward(Remote<ServerAddr>, Metadata),
+}
+
+/// Wraps errors encountered in this module.
 #[derive(Debug, thiserror::Error)]
 #[error("concrete service {addr}: {source}")]
 pub struct ConcreteError {
-    addr: ConcreteAddr,
+    addr: NameAddr,
     #[source]
     source: Error,
 }
 
+/// Inner stack target type.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Endpoint<T> {
+    addr: Remote<ServerAddr>,
+    is_local: bool,
+    metadata: Metadata,
+    parent: T,
+}
+
+/// A target configuring a load balancer stack.
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct Balance<T> {
+    addr: NameAddr,
+    ewma: balance::EwmaConfig,
+    parent: T,
+}
+
 // === impl Outbound ===
 
 impl<C> Outbound<C> {
     /// Builds a [`svc::NewService`] stack that builds buffered opaque TCP load
-    /// balancer services for `Concrete` targets.
+    /// balancer services for `T`-typed concrete targets.
     ///
     /// When a balancer has no available inner services, it goes into
     /// 'failfast'. While in failfast, buffered requests are failed and the
     /// service becomes unavailable so callers may choose alternate concrete
     /// services.
     //
     // TODO(ver) make the outer target type generic/parameterized.
```

Comment:
i feel like this doc comment should maybe be updated to also indicate that this is the stack that determines whether to build a TCP load-balancer stack or forward directly to an endpoint?